src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_gligen_text_image.py

```diff
@@ -0,0 +1,979 @@
+                )
+
+    # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.prepare_latents
+    def prepare_latents(self, batch_size, num_channels_latents, height, width, dtype, device, generator, latents=None):
+        shape = (batch_size, num_channels_latents, height // self.vae_scale_factor, width // self.vae_scale_factor)
+        if isinstance(generator, list) and len(generator) != batch_size:
+            raise ValueError(
+                f"You have passed a list of generators of length {len(generator)}, but requested an effective batch"
+                f" size of {batch_size}. Make sure the batch size matches the length of the generators."
+            )
+
+        if latents is None:
+            latents = randn_tensor(shape, generator=generator, device=device, dtype=dtype)
+        else:
+            latents = latents.to(device)
+
+        # scale the initial noise by the standard deviation required by the scheduler
+        latents = latents * self.scheduler.init_noise_sigma
+        return latents
+
+    def enable_fuser(self, enabled=True):
+        for module in self.unet.modules():
+            if type(module) is GatedSelfAttentionDense:
+                module.enabled = enabled
+
+    def draw_inpaint_mask_from_boxes(self, boxes, size):
+        inpaint_mask = torch.ones(size[0], size[1])
+        for box in boxes:
+            x0, x1 = box[0] * size[0], box[2] * size[0]
+            y0, y1 = box[1] * size[1], box[3] * size[1]
+            inpaint_mask[int(y0) : int(y1), int(x0) : int(x1)] = 0
+        return inpaint_mask
+
+    def crop(self, im, new_width, new_height):
+        width, height = im.size
+        left = (width - new_width) / 2
+        top = (height - new_height) / 2
+        right = (width + new_width) / 2
+        bottom = (height + new_height) / 2
+        return im.crop((left, top, right, bottom))
+
+    def target_size_center_crop(self, im, new_hw):
+        width, height = im.size
+        if width != height:
+            im = self.crop(im, min(height, width), min(height, width))
+        return im.resize((new_hw, new_hw), PIL.Image.LANCZOS)
+
+    def project(self, x, hidden_size, device):
+        """
+        x (Batch*768) should be the penultimate feature of CLIP (before projection) projection_matrix (768*768) is the
+        CLIP projection matrix, which should be weight.data of Linear layer defined in CLIP (out_dim, in_dim), thus we
+        need to apply transpose below. this function will return the CLIP feature (without normalziation)
+        """
+        projection_matrix = hf_hub_download(repo_id="anhnct/Gligen_Text_Image", filename="projection_matrix")
+        projection_matrix = torch.load(projection_matrix).type(self.image_encoder.dtype).to(device).T
+        if hidden_size != projection_matrix.shape[0]:
+            return x
+        else:
+            return x @ torch.transpose(projection_matrix, 0, 1)
```

Comment:
Not too happy about doing it this way. Can't we leverage the `hidden_states` of the CLIP text model and index the required features from there? @patrickvonplaten WDYT?

Reply:
In the official GLIGEN repository, they utilize the projection_matrix. In my implementation, I attempted to solely use the hidden_state of the CLIP model without employing the projection_matrix. However, upon comparing the outcomes, not utilizing the projection_matrix led to discrepancies in the results when compared to the official repository. If I use the projection_matrix here, the results produced in the original repository and the Diffusers implementation are the same.(Here, I used the same seed during inference)

Reply:
Is this projection matrix not a part of the CLIP text model? If so, I think we should be able to leverage the matrix, no?

Reply:
They don't mention where the projection_matrix come from. They just provide a torch tensor of it.

Reply:
![image](https://github.com/huggingface/diffusers/assets/52446297/e1b6a410-29fd-4631-8249-e5066aeabb9c)
I think projection_matrix is not a part of CLIP model. They have it after training, i think so.

Reply:
If that is the case, I think it makes sense to implement that as a separate `nn.Module` and keep that under the pipeline directory. We should then consider making that a part of the pipeline init i.e., a component of the pipeline (just like the UNet, text encoder, etc.). Maybe `projection` component. 

WDYT?

Reply:
I think that's quite reasonable, and I will proceed with implementing this idea.